### ISSUE TITLE
Selecting fruit that is not citrus does not work.

### ISSUE THREAD

**Issue by hanjoosten** [AmpersandTarski/ampersand#237]
Given the well-known fruit example, 

CLASSIFY Citrus ISA Fruit

In an interface of concept Fruit, I want to show iff it is not Citrus:

BOX 
[ "fruit" : I
, "non-citrus fruit" : I[Fruit] /\ -I[Citrus]
]

In the sql, I get:

    '/* case: ECpl (EDcI Citrus) */
    select
    distinct
    "concept0"."Citrus" as src, "concept1"."Citrus" as tgt
    from "Fruit" as "concept0",
             "Fruit" as "concept1"
    where "concept0"."Citrus" <> "concept1"."Citrus"'

This is obviously wrong, because it does not select the fruits that are not citrus.

The expression in the first line comes from the normalizer. It doesn't tell anything about the largest concept. That is required, because otherwise the SQL generator has to guess.

**Comment by hanjoosten** [comment]
@stefjoosten: What expression should the normalizer give to the SQL generator in this case?

**Comment by stefjoosten** [comment]
#Preparation
I am experimenting in `git\ampersand-models\Tests\ShouldSucceed\Try53.adl`. In this way, the solution that we find might be fit to run along in the Sentinel.

# Analysis
The normalizer performs the following proof before generating code:

                            // Normalization steps:
                            //     I{-Eps-}[Citrus];I[Fruit];I{-Eps-}[Citrus] /\ -I[Citrus]
                            // <=> { I;x = x }
                            //     I{-Eps-}[Citrus];I{-Eps-}[Citrus] /\ -I[Citrus]
                            // <=>
                            //     I{-Eps-}[Citrus] /\ -I[Citrus]
                            // <=> { source and target are equal to Citrus, so I{-Eps-}[Citrus]=I[Citrus] }
                            //     I[Citrus] /\ -I[Citrus]
                            // <=> { I[Citrus] /\ -I[Citrus] = V- }
                            //     -V[Citrus*Citrus]
                            //
                            // Normalized interface expression (== expressionSQL): -V[Citrus*Citrus]

Obviously, the steps of the proof are correct, but the starting point is not.
The first expression should not have ` I{-Eps-}[Citrus]` on the left.

**Comment by sjcjoosten** [comment]
This is a bug in the type system. Here is the work-around:

``
BOX 
[ "fruit" : I
, "non-citrus fruit" : I[Fruit] /\ V[Fruit*Fruit]-I[Citrus]
]
``

The desired behavior will be a type error, stating that we do not know the type of `-I[Citrus]`: it could either be `[Citrus*Citrus]` or `[Fruit*Fruit]` or `[Fruit*Citrus]` or `[Citrus*Fruit]`. You may move Try53 to ShouldFail. I will fix this with a (hopefully slightly) more cryptic error message, and create a separate ticket to improve that message. Improving the message is more involved, unfortunately.

**Comment by stefjoosten** [comment]
@hanjoosten Thanks for finding this bug.
Please,
- Close this issue if you are satisfied with the solution, or
- Close this issue if you are not satisfied with the solution.

:smile:

**Comment by hanjoosten** [comment]
I can only tell if I am satisfied with this solution if I know how it is ensured that @sjcjoosten is not going to forget about this. Is there another ticket that guarantees that? If so, which one? Of course, I will close this ticket as soon as I know my decision on the satisfaction about it. :smile_cat:

(allthough, in the case that I would not be satisfied, I might directly reopen it again)

**Comment by sjcjoosten** [comment]
Hi Han, I am in a bit of a difficult situation right now: I cannot compile anything from the master branch on my computer, so I do not know what the current error message is on Try53.adl. The cause of this problem lies with ghc 6.8.3 being incompatible with the current version of my OS (due to a security update in my OS, which I do not wish to disable), so I will not fix this but just wait for ampersand to use 7.10.2, which does work.

Hopefully @hanjoosten can describing the current output of Try53.adl, which is probably a very bad error message, so I can make a ticket for it. I'm closing this one, but you're free to reopen of course :smile_cat:

**Comment by hanjoosten** [comment]
I reopened this issue. I did not find any Try53.adl in shouldfail. So I put it there myself. 

Running ampersand 62e159f, it does not fail. It passes. However, it should fail, as the above analisys suggests. 

@sjcjoosten, you could (if you make Try53.adl fail, and as long as you are not able to check what error message is produced) temporarily copy Try53.adl to shouldsucceed. That should make sentinel show the error message. Whenever you are content, you could remove it  again.

**Comment by stefjoosten** [comment]
That's my fault. I could not push `Ticket237.adl` to ShouldFail, because I got in a mess with Git and insufficient time to solve it. That explains its absence. Sorry for that. Nevertheless, I do want a file `Tests\ShouldFail\Ticket237.adl` that tests this issue in the sentinel. Once it is in place, Han can close this issue as far as I'm concerned.

**Comment by stefjoosten** [comment]
I have included two tests in the Sentinel:
1. `Tests/ShouldSucceed/Ticket237.adl`, which proves that the interpretation of `I[Fruit]-I[Citrus]` corresponds with `I[Fruit]-(V[Fruit]-I[Citrus])` for a well-chosen population.
2. `Tests/ShouldFail/Ticket237.adl`, which proves that `I[Fruit] /\ -I[Citrus]` is rejected by the type checker due to ambiguity. The error message is:

    An ambiguity arises in type checking. Be more specific by binding the Src of the expression I{-Eps-}[Citrus];I[Fruit];I{-Eps-}[Citrus] /\ -I[Citrus]. 
    You could add more types inside the expression, or just write (I{-Eps-}[Citrus];I[Fruit];I{-Eps-}[Citrus] /\ -I[Citrus]) /\ V[Citrus*Citrus]. 
    Error at symbol () in file /home/sentinel/git/ampersand-models/Tests/ShouldFail/Ticket237.adl at line 6 : 29
